services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -496,6 +564,143 @@ public Response requestCredential(CredentialRequest credentialRequestVO) {
         return Response.ok().entity(responseVO).build();
     }
 
+    /**
+     * Decrypts a JWE-encoded Credential Request and validates it against metadata.
+     *
+     * @param jweString The JWE compact serialization
+     * @param metadata The CredentialRequestEncryptionMetadata
+     * @return The parsed CredentialRequest
+     * @throws JWEException If decryption or validation fails
+     */
+    private CredentialRequest decryptCredentialRequest(String jweString, CredentialRequestEncryptionMetadata metadata) throws Exception {
+        JWE jwe = new JWE(jweString);
+        JWEHeader header = (JWEHeader) jwe.getHeader();
+
+        // Validate alg and enc against supported values
+        String enc = header.getEncryptionAlgorithm();
+        if (!metadata.getEncValuesSupported().contains(enc)) {
+            String errorMessage = String.format("Unsupported content encryption algorithm: enc=%s", enc);
+            LOGGER.debugf(errorMessage);
+            throw new JWEException(String.valueOf(ErrorType.INVALID_ENCRYPTION_PARAMETERS));
+        }
+
+        // Handle compression if present
+        String zip = header.getCompressionAlgorithm();
+        if (zip != null && (metadata.getZipValuesSupported() == null || !metadata.getZipValuesSupported().contains(zip))) {
+            String errorMessage = String.format("Unsupported compression algorithm: zip=%s", zip);
+            LOGGER.debugf(errorMessage);
+            throw new JWEException(String.valueOf(ErrorType.INVALID_ENCRYPTION_PARAMETERS));
+        }
+
+        // Get a private key from KeyManager based on kid
+        String kid = header.getKeyId();
+        if (kid == null) {
+            throw new JWEException("Missing kid in JWE header");
+        }
+
+        RealmModel realm = session.getContext().getRealm();
+        KeyManager keyManager = session.keys();
+        List<KeyWrapper> matchingKeys = keyManager.getKeysStream(realm)
+                .filter(key -> KeyUse.ENC.equals(key.getUse()) && kid.equals(key.getKid()))
+                .collect(Collectors.toList());
+
+        if (matchingKeys.isEmpty()) {
+            throw new JWEException("No encryption key found for kid: " + kid);
+        }
+        if (matchingKeys.size() > 1) {
+            throw new JWEException("Multiple encryption keys found for kid: " + kid);
+        }
+        KeyWrapper keyWrapper = matchingKeys.get(0);
+
+        // Set the decryption key
+        jwe.getKeyStorage().setDecryptionKey(keyWrapper.getPrivateKey());
+
+        // Decrypt the JWE
+        try {
+            jwe.verifyAndDecodeJwe();
+        } catch (JWEException e) {
+            throw new JWEException("Failed to decrypt JWE: " + e.getMessage());
+        }
+
+        // Handle decompression if zip=DEF
+        byte[] content = jwe.getContent();
+        if (zip != null && "DEF".equals(zip)) {
```

Comment:
This is a superfluous double-check. The decompress method also checks zip again.